bigbang/utils.py

```diff
@@ -1,13 +1,173 @@
+                return "RequestException"
+        else:
+            sauce = session.get(url)
+            soup = BeautifulSoup(sauce.text, "html.parser")
+        return soup
+    except requests.exceptions.RequestException as e:
+        if "A2=" in url:
+            # if URL of mboxMessage
+            logger.info(f"{e} for {url}.")
+            return "RequestException"
+        else:
+            SystemExit()
+
+
+def set_website_preference_for_header(
+    url_pref: str,
+    session: requests.Session,
+) -> requests.Session:
+    """
+    Set the 'Email Headers' of the 'Archive Preferences' for the auth session
+    to 'Show All Headers'. Otherwise only a restricted list of header fields is
+    shown.
+    """
+    url_archpref = url_pref + "&TAB=2"
+    payload = {
+        "Email Headers": "b",
+    }
+    session.post(url_archpref, data=payload)
+    return session
+
+
+def get_auth_session(
+    url_login: str, username: str, password: str
+) -> requests.Session:
+    """
+    Create AuthSession.
+
+    There are three ways to create an AuthSession:
+        - parse username & password directly into method
+        - create a /bigbang/config/authentication.yaml file that contains keys
+        - type then into terminal when the method 'get_login_from_terminal'
+            is raised
+    """
+    if os.path.isfile(filepath_auth):
+        # read from /config/authentication.yaml
+        with open(filepath_auth, "r") as stream:
+            auth_key = yaml.safe_load(stream)
+        username = auth_key["username"]
+        password = auth_key["password"]
+    else:
+        # ask user for login keys
+        username, password = get_login_from_terminal(username, password)
+
+    if username is None or password is None:
+        # continue without authentication
+        return None
+    else:
+        # Start the AuthSession
+        session = requests.Session()
+        # Create the payload
+        payload = {
+            "LOGIN1": "",
+            "Y": username,
+            "p": password,
+            "X": "",
+        }
+        # Post the payload to the site to log in
+        session.post(url_login, data=payload)
+        return session
+
+
+def get_login_from_terminal(
+    username: Union[str, None],
+    password: Union[str, None],
+    file_auth: str = directory_project + "/config/authentication.yaml",
+) -> Tuple[Union[str, None]]:
+    """
+    Get login key from user during run time if 'username' and/or 'password' is 'None'.
+    Return 'None' if no reply within 15 sec.
+    """
+    if username is None or password is None:
+        record = True
+    else:
+        record = False
+    if username is None:
+        username = ask_for_input("Enter your Email: ")
+    if password is None:
+        password = ask_for_input("Enter your Password: ")
+    if record and isinstance(username, str) and isinstance(password, str):
+        loginkey_to_file(username, password, file_auth)
+    return username, password
```

Comment:
Looks like many of these new functions have to do specifically with managing interactions with a remote website that requires login.

Maybe they should be in a separate submodule with a more descriptive name than utils.py

Reply:
I now moved ingress related functions from `bigbang/utils.py` to `bigbang/ingress/utils.py` in this PR